methods/enumerables.rb

```diff
@@ -0,0 +1,154 @@
+module Enumerable
+  def range_return_array(arr)
+    if self.class == Range
+      self
+    else
+      arr
+    end
+  end
+
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    arr = to_a
+    arr.size.times do |index|
+      yield arr[index]
+    end
+    range_return_array(arr)
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each) unless block_given?
+
+    arr = to_a
+    arr.size.times do |index|
+      yield arr[index], index
+    end
+    range_return_array(arr)
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = to_a
+    answer = []
+    arr.my_each do |item|
+      answer.push(item) if yield item
+    end
+    answer
+  end
+
+  def my_all?(arg = nil)
+    arr = to_a
+
+    if block_given?
+      arr.my_each { |item| return false unless yield item }
+      true
+    elsif !block_given? && !arg
+      arr.my_each { |item| return false unless item }
+      true
+    elsif arg.class == Class
+      arr.my_each { |item| return false unless item.is_a? arg }
+      true
+    elsif arg.class == Regexp
+      arr.my_each { |item| return false unless item.class =~ arg }
+      true
+    elsif arg.class != Regexp && arg.class != Class
+      arr.my_each { |item| return false unless item == arg }
+      true
+    end
+  end
+
+  def my_any?(arg = nil)
+    arr = to_a
+
+    if block_given?
+      arr.my_each { |item| return true if yield item }
+      false
+    elsif !block_given? && !arg
+      arr.my_each { |item| return true if item }
+      false
+    elsif arg.class == Class
+      arr.my_each { |item| return true if item.is_a? arg }
+      false
+    elsif arg.class == Regexp
+      arr.my_each { |item| return true if item.class =~ arg }
+      false
+    elsif arg.class != Regexp && arg.class != Class
+      arr.my_each { |item| return true if item == arg }
+      false
+    end
+  end
+
+  def my_none?(arg = nil, &block)
+    !my_any?(arg, &block)
+  end
+
+  def my_count(arg = nil)
+    count = 0
+    if arg
+      my_each do |item|
+        count += 1 if item == arg
+      end
+    elsif block_given?
+      my_each do |item|
+        count += 1 if (yield item) == true
+      end
+    else
+      my_each do
+        count += 1
+      end
+    end
+    count
+  end
+
+  def my_map(arg = nil)
+    return to_enum if !block_given? && !arg
+
+    my_each_with_index do |item, i|
+      self[i] = if arg
+                  arg.call(item)
+                else
+                  yield item
+                end
+    end
+    self
+  end
```

Comment:
- [ ] it returns a new array with the results of running a given block once for every element in enum.

```Ruby
array.my_map { |num| num < 10 }
     
     NoMethodError:
       undefined method `<' for true:TrueClass
```

- [ ] it returns a new array with the results of running a given block once for every element in enum.

```Ruby
Failure/Error: block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
     
     NoMethodError:
       undefined method `<' for true:TrueClass
```

- [ ] it  returns a new array with the results of running a given block once for every element in enum::range

```Ruby
  range.my_map(&block)

     NoMethodError:
       undefined method `[]=' for 5..50:Range
```

- [ ] it accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

```Ruby
  Failure/Error: my_proc = proc { |num| num > 10 } 

     NoMethodError:
       undefined method `>' for false:FalseClass

```

- [ ] it executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

```Ruby
  Failure/Error: my_proc = proc { |num| num > 10 }

     NoMethodError:
       undefined method `>' for false:FalseClass
```

- [ ] it does not mutate the original array

```Ruby
  array.my_select { |num| num + 1 }
```

expected: `[2, 7, 0, 4, 1, 7, 2, 1, 1, 6, 2, 0, 4, 6, 1, 6, 4, 7, 3, 8, 8, 6, 1, 5, 0, 1, 4, 7, 8, 6, 0, 1, 7, 6..., 3, 8, 7, 8, 6, 8, 3, 3, 1, 0, 1, 4, 2, 2, 1, 4, 2, 3, 8, 8, 7, 5, 1, 1, 5, 0, 6, 8, 5, 4, 6, 8, 7]`
got: `[3, 8, 1, 5, 2, 8, 3, 2, 2, 7, 3, 1, 5, 7, 2, 7, 5, 8, 4, 9, 9, 7, 2, 6, 1, 2, 5, 8, 9, 7, 1, 2, 8, 7..., 4, 9, 8, 9, 7, 9, 4, 4, 2, 1, 2, 5, 3, 3, 2, 5, 3, 4, 9, 9, 8, 6, 2, 2, 6, 1, 7, 9, 6, 5, 7, 9, 8]`